### PR TITLE
🌱 cache: introduce the replication controller

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
The new controller copies objects of defined resources that have the `internal.sharding.kcp.dev/replicate` annotation to the cache server.
The replicated object will be placed under the same cluster as the original object.
In addition to that, all replicated objects will be placed under the shard taken from the shardName argument.
For example: shards/{shardName}/clusters/{clusterName}/apis/apis.kcp.dev/v1alpha1/apiexports

As of today the controller replicates the ApiExport resources.
The reconcileAPIExports function handles the following cases:
1. creation of the object in the cache server
   happens when the cached object is not found in c.cacheApiExportsIndexer
2. deletion of the object from the cache server
   happens when either of the following is true:
   - the original/local object was removed
   - the original/local object was not found in c.localApiExport
3. modification of the object to match the original object
   happens when either of the following is true:
    - the original object's metadata doesn't match the cached one.
    - the original object's spec doesn't match the cached one.
    - the original object's status doesn't match the cached one.

TODO:
 - in a follow-up PR I am going to add `APIResourceSchema` to this controller

## Related issue(s)

https://github.com/kcp-dev/kcp/issues/342

### CODE REVIEW

File: pkg/reconciler/replication/indexers.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (sttts) [inline pkg/reconciler/replication/indexers.go:1]: pkg/reconciler/cache/replication ?

File: pkg/reconciler/replication/indexers.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+const (
+	// ByShardAndLogicalCluster is the name for the index that indexes by an object's shard and logical cluster.
+	ByShardAndLogicalCluster = "kcp-byShardAndLogicalCluster"
+
+	// ByShardAndLogicalClusterAndNamespace is the name for the index that indexes by an object's shard and logical cluster and namespace.
+	ByShardAndLogicalClusterAndNamespace = "kcp-byShardAndLogicalClusterAndNamespace"
+)
+
+// IndexByShardAndLogicalCluster is an index function that indexes by an object's shard and logical cluster.
+func IndexByShardAndLogicalCluster(obj interface{}) ([]string, error) {
```
Comment (sttts) [inline pkg/reconciler/replication/indexers.go:38]: capital?

File: pkg/reconciler/replication/replication_reconcile_apiexports.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	if !statusChanged {
+		return nil
+	}
+	_, err = c.cacheKcpClusterClient.Cluster(logicalcluster.From(localApiExport)).ApisV1alpha1().APIExports().UpdateStatus(ctx, copyCacheApiExport, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *controller) retrieveApiExport(apiExportName, clusterName string) (*apisv1alpha1.APIExport, error) {
+	var cacheApiExport *apisv1alpha1.APIExport
+	cacheApiExports, err := c.cacheApiExportsIndexer.ByIndex(ByShardAndLogicalCluster, toShardAwareKey(c.shardName, clusterName))
```
Comment (stevekuznetsov) [inline pkg/reconciler/replication/replication_reconcile_apiexports.go:103]: ah ack you do share the formatter, but if you export one, let's export the other

File: pkg/reconciler/replication/replication_reconcile_apiexports.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+//     happens when either of the following is true:
+//     - the original/local object was removed
+//     - the original/local object was not found in c.localApiExport
+//  3. modification of the object to match the original object
+//     happens when either of the following is true:
+//      - the original object's metadata doesn't match the cached one.
+//        as of today we compare only the annotations and the labels
+//      - the original object's spec doesn't match the cached one.
+//      - the original object's status doesn't match the cached one.
+func (c *controller) reconcileAPIExports(ctx context.Context, key string) error {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/replication/replication_reconcile_apiexports.go:48]: oh, yes, we could, only if all types have the same structure i.e. `meta, spec and status`. At least the two types I want to add have the same structure. I will look into that.

File: pkg/reconciler/replication/replication_reconcile_apiexports.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return err
+	}
+	if localApiExport.DeletionTimestamp != nil {
+		return c.handleApiExportDeletion(ctx, localApiExport, cacheApiExport)
+	}
+
+	if cacheApiExport == nil {
+		copyLocalApiExport := localApiExport.DeepCopy()
+		copyLocalApiExport.ResourceVersion = ""
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/replication/replication_reconcile_apiexports.go:72]: I'll add a `TODO` but I think the RV we want to keep track of will end up in an annotation.

File: pkg/reconciler/replication/replication_reconcile_apiexports.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	objectMetaChanged := ensureObjectMeta(&copyCacheApiExport.ObjectMeta, localApiExport.ObjectMeta)
+	specChanged := ensureApiExportSpec(&copyCacheApiExport.Spec, localApiExport.Spec)
+	statusChanged := ensureApiExportStatus(&copyCacheApiExport.Status, localApiExport.Status)
+
+	if !specChanged && !objectMetaChanged && !statusChanged {
+		return nil
+	}
+
+	if specChanged || objectMetaChanged {
+		_, err = c.cacheKcpClusterClient.Cluster(logicalcluster.From(localApiExport)).ApisV1alpha1().APIExports().Update(ctx, copyCacheApiExport, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/replication/replication_reconcile_apiexports.go:90]: we reconcile updates to meta and status first and then to the status, does it make sense ?

File: pkg/reconciler/replication/replication_reconcile.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		// when the kind and apiversion are not set
+		unstructuredCacheApiExport.SetKind("APIExport")
+		unstructuredCacheApiExport.SetAPIVersion(gvr.GroupVersion().String())
+	}
+	if localApiExport != nil {
+		unstructuredLocalApiExport, err = toUnstructured(localApiExport)
+		if err != nil {
+			return err
+		}
+		// TODO: for some reason type data is empty, creating an obj with the dynamic client fails
```
Comment (p0lyn0mial) [inline pkg/reconciler/replication/replication_reconcile.go:105]: does anyone know why it doesn't work?

File: pkg/reconciler/replication/replication_reconcile_apiexports.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/replication/replication_reconcile_apiexports.go]: updated the controller to operate on `*unstructured.Unstructured`, PTAL.

File: pkg/reconciler/replication/replication_reconcile_apiexports.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+//     - the original/local object was removed
+//     - the original/local object was not found in c.localApiExport
+//  3. modification of the object to match the original object
+//     happens when either of the following is true:
+//      - the original object's metadata doesn't match the cached one.
+//        as of today we compare only the annotations and the labels
+//      - the original object's spec doesn't match the cached one.
+//      - the original object's status doesn't match the cached one.
+func (c *controller) reconcileAPIExports(ctx context.Context, key string) error {
+	localApiExport, err := c.localApiExportLister.Get(key)
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/replication/replication_reconcile_apiexports.go]: I think this key is correct, why do you think it is incorrect? 
BTW: I think that the key format is `cluster|namespace/name`

File: pkg/reconciler/cache/replication/replication_controller.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	// ControllerName hold this controller name.
+	ControllerName = "kcp-replication-controller"
+
+	// AnnotationKey is the name of the annotation key used to mark an object for replication.
+	AnnotationKey = "sharding.kcp.dev/replicate"
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_controller.go:46]: `internal.sharding.kcp.dev/replicate`. This is not an API.

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+//  1. creation of the object in the cache server when the cached object is not found in c.localApiExportLister
+//  2. deletion of the object from the cache server when the original/local object was removed OR was not found in c.localApiExportLister
+//  3. modification of the cached object to match the original one when meta.annotations, meta.labels, spec or status are different
+func (c *controller) reconcileAPIExports(ctx context.Context, key string, gvr schema.GroupVersionResource) error {
+	var cacheApiExport *apisv1alpha1.APIExport
+	var localApiExport *apisv1alpha1.APIExport
+	var cluster logicalcluster.Name
+	var err error
+
+	localApiExport, err = c.localApiExportLister.Get(key)
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:59]:
```suggestion

	var err error
	localApiExport, err = c.localApiExportLister.Get(key)
```

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	if errors.IsNotFound(err) {
+		var apiExportName string
+		cluster, _, apiExportName, err = kcpcache.SplitMetaClusterNamespaceKey(key)
+		if err != nil {
+			return err
+		}
+		cacheApiExport, err = c.retrieveApiExport(apiExportName, cluster.String())
+		if err != nil {
+			return err
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:71]: why couldn't this be NotFound ?

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+		}
+		// issue a live GET to make sure the localApiExport was removed
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:73]: what's the logic here? Why can a local informer be stale and why does it matter?

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+		unstructuredLiveLocalApiExport, err := c.dynamicLocalClient.Cluster(cluster).Resource(gvr).Get(ctx, apiExportName, metav1.GetOptions{})
+		if err == nil {
+			return fmt.Errorf("the informer used by this controller is stale, the following APIExport was found on the local server: %s/%s/%s but was missing in the informer", cluster, unstructuredLiveLocalApiExport.GetNamespace(), unstructuredLiveLocalApiExport.GetName())
+		}
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	if cacheApiExport == nil {
+		cacheApiExport, err = c.retrieveApiExport(localApiExport.Name, logicalcluster.From(localApiExport).String())
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:83]: why don't we do this as first step? Then we don't need two calls here.

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+	var unstructuredCacheApiExport *unstructured.Unstructured
+	var unstructuredLocalApiExport *unstructured.Unstructured
+	if cacheApiExport != nil {
+		unstructuredCacheApiExport, err = toUnstructured(cacheApiExport)
+		if err != nil {
+			return err
+		}
+		// TODO: for some reason type data is empty, creating an obj with the dynamic client fails
+		// when the kind and apiversion are not set
+		unstructuredCacheApiExport.SetKind("APIExport")
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:98]: that's normal. The dynamic client cannot know what to fill in here.

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	if cluster.Empty() && localApiExport != nil {
+		cluster = logicalcluster.From(localApiExport)
+	}
+
+	return c.reconcileUnstructuredObject(ctx, cluster, &gvr, unstructuredCacheApiExport, unstructuredLocalApiExport)
+}
+
+func (c *controller) retrieveApiExport(apiExportName, clusterName string) (*apisv1alpha1.APIExport, error) {
+	var cacheApiExport *apisv1alpha1.APIExport
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:119]: why is this needed? Just return below in the loop.

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+	cacheApiExports, err := c.cacheApiExportsIndexer.ByIndex(ByShardAndLogicalCluster, toShardAwareKey(c.shardName, clusterName))
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:120]: why not a shard,lcluster,name index?

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 12 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:73]: an informer can be stale but we should ensure the original/object was removed before removing its cached counterpart, no?

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 16 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:120]: yes, we could to that, it won't have a huge footprint on memory consumption.

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 14 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:98]: aren't those fields persisted?
does the client-go auto-fills them?

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 15 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:119]: yes, should be removed, thx.

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+	if specChanged || metaChanged {
+		_, err := c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Namespace(unstructuredCacheObject.GetNamespace()).Update(ctx, unstructuredCacheObject, metav1.UpdateOptions{})
+		return err
+	}
+	if !statusChanged {
+		return nil
+	}
+	_, err = c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Namespace(unstructuredCacheObject.GetNamespace()).UpdateStatus(ctx, unstructuredCacheObject, metav1.UpdateOptions{})
+	return err
+}
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:89]: these subresource distinctions surprise me. Why not disable the status subresource in the cache? Then we can skip all of this dance.

Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:89]: I can look into that but I think it comes from the extension server we use. Might require yet another patch in the fork.

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+func (c *controller) handleUnstructuredObjectDeletion(ctx context.Context, cluster logicalcluster.Name, gvr *schema.GroupVersionResource, unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) error {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:91]: I can look into that but I think it comes from the extension server we use. Might require yet another patch in the fork.

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 11 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:71]: indexers don't return `NotFound` only an empty dataset when not found, but I can change that.

File: pkg/reconciler/cache/replication/replication_controller.go (hunk 9 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_controller.go]: done

File: pkg/reconciler/cache/replication/indexers.go [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+	annotations := a.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	shardName := annotations[genericapirequest.AnnotationKey]
+
+	return []string{toShardAwareKey(shardName, clusters.ToClusterAwareKey(logicalcluster.From(a), a.GetNamespace()))}, nil
+}
+
+func toShardAwareKey(shardName, key string) string {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/indexers.go]: done.

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 13 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go]: done

File: pkg/reconciler/cache/replication/indexers.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// ByShardAndLogicalClusterAndNamespaceAndName is the name for the index that indexes by an object's shard and logical cluster, namespace and name
+	ByShardAndLogicalClusterAndNamespaceAndName = "kcp-byShardAndLogicalClusterAndNamespaceAndName"
+)
+
+// IndexByShardAndLogicalClusterAndNamespace is an index function that indexes by an object's shard and logical cluster, namespace and name
+func IndexByShardAndLogicalClusterAndNamespace(obj interface{}) ([]string, error) {
```
Comment (p0lyn0mial) [inline pkg/reconciler/cache/replication/indexers.go:34]: I've decided to provide only this indexer

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go (hunk 17 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:89]: @sttts you were right, turning off subresources did the trick :) PTAL - https://github.com/kcp-dev/kcp/pull/2063

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:91]: addressed in https://github.com/kcp-dev/kubernetes/pull/102

**Comment by p0lyn0mial** [comment]
@sttts @stevekuznetsov updated the PR, it is ready for another pass.

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go (hunk 17 above)
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:85]: 🥳

File: pkg/reconciler/cache/replication/replication_controller.go [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+	if err := cacheApiExportInformer.Informer().AddIndexers(cache.Indexers{
+		ByShardAndLogicalClusterAndNamespaceAndName: IndexByShardAndLogicalClusterAndNamespace,
+	}); err != nil {
+		return nil, err
+	}
+	c.cacheApiExportsIndexer = cacheApiExportInformer.Informer().GetIndexer()
+
+	localApiExportInformer.Informer().AddEventHandler(c.apiExportInformerEventHandler())
+	cacheApiExportInformer.Informer().AddEventHandler(c.apiExportInformerEventHandler())
```
Comment (p0lyn0mial) [inline pkg/reconciler/cache/replication/replication_controller.go:81]: In the future, we might want to find a more efficient way of storing and comparing data from the cache. Using an informer double-down memory.

Comment (sttts) [inline pkg/reconciler/cache/replication/replication_controller.go:81]: am curious, where `cacheApiExportInformer` is set up, we don't wait for sync, do we? So we have to wait here internally. (have not read much yet, maybe you do that somewhere)

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+//  1. creation of the object in the cache server when the cached object is not found in c.localApiExportLister
+//  2. deletion of the object from the cache server when the original/local object was removed OR was not found in c.localApiExportLister
+//  3. modification of the cached object to match the original one when meta.annotations, meta.labels, spec or status are different
+func (c *controller) reconcileAPIExports(ctx context.Context, key string, gvr schema.GroupVersionResource) error {
+	var cacheApiExport *apisv1alpha1.APIExport
+	var localApiExport *apisv1alpha1.APIExport
+	var cluster logicalcluster.Name
+	var namespace string
+	var apiExportName string
+	var err error
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile.go:59]: why all these vars here? Just do `:= below`

Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:59]: C77 style :D

Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:59]: I was near, it was 78: https://en.wikipedia.org/wiki/C_(programming_language)#K&R_C

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 14 above)
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:90]: dynamic client yes, typed clients no afaik

File: pkg/reconciler/cache/replication/replication_controller.go (hunk 21 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_controller.go:81]: the controller hasn't been wired yet. we will wait in the `installXYZ` method, before we start the controller.

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 22 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:59]: :)

Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:59]: addressed in https://github.com/kcp-dev/kcp/pull/2090/commits/245aada5b5c970f77c0c4435548cff421aa371d2

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+	}
+	if unstructuredCacheObject.GetDeletionTimestamp() == nil {
+		return c.dynamicCacheClient.Cluster(cluster).Resource(*gvr).Delete(ctx, unstructuredCacheObject.GetName(), metav1.DeleteOptions{})
+	}
+	return nil
+}
+
+// ensureUnstructuredMeta changes unstructuredCacheObject's metadata to match unstructuredLocalObject's metadata.
+// returns true when the unstructuredCacheObject was updated.
+func ensureUnstructuredMeta(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:99]: why doesn't this do more than just the few fields?

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+	unstructuredCacheObject.SetLabels(cacheObjectNewLabels)
+
+	return true, nil
+}
+
+// ensureUnstructuredSpec changes unstructuredCacheObject's spec to match unstructuredLocalObject's spec.
+// returns true when the unstructuredCacheObject was updated.
+//
+// note that the spec field is not required, in that case this function is no-op
+func ensureUnstructuredSpec(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:136]: am confused that we even distinguish between spec, status, ... Why not just update everything?

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+		return false, nil // no-op
+	}
+	if !hasUnstructuredLocalObjectMetaRaw {
+		unstructured.RemoveNestedField(unstructuredCacheObject.Object, "metadata")
+		return true, nil
+	}
+
+	// before we can compare the cache object we need to
+	// store, remove and then bring back fields that are unique only to the cache object
+	if unstructuredCacheObjectResourceVersion, hasUnstructuredCacheObjectResourceVersion := unstructuredCacheObjectMeta["resourceVersion"]; hasUnstructuredCacheObjectResourceVersion {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:120]: shorter var names. Here is so much context that we can have `rv, found :=`

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+		unstructured.RemoveNestedField(unstructuredCacheObjectMeta, "resourceVersion")
+		defer func() {
+			if err == nil {
+				err = unstructured.SetNestedField(unstructuredCacheObject.Object, unstructuredCacheObjectResourceVersion, "metadata", "resourceVersion")
+			}
+		}()
+	}
+	if unstructuredCacheObjectAnnotationsRaw, hasUnstructuredCacheObjectAnnotations := unstructuredCacheObjectMeta["annotations"]; hasUnstructuredCacheObjectAnnotations {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:128]: same here, short vars

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+		unstructuredCacheObjectAnnotations, ok := unstructuredCacheObjectAnnotationsRaw.(map[string]interface{})
+		if !ok {
+			return false, fmt.Errorf("metadata.annotations field of unstructuredCacheObject is of the type %T, expected map[string]interface{}", unstructuredCacheObjectAnnotationsRaw)
+		}
+		if shardName, foundShardName := unstructuredCacheObjectAnnotations[genericrequest.AnnotationKey]; foundShardName {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:133]: `shard, found`

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	unstructuredLocalObjectMeta, ok := unstructuredLocalObjectMetaRaw.(map[string]interface{})
+	if !ok {
+		return false, fmt.Errorf("metadata field of unstructuredLocalObjectMeta is of the type %T, expected map[string]interface{}", unstructuredLocalObjectMetaRaw)
+	}
+	if !hasUnstructuredLocalObjectMetaRaw && !hasUnstructuredCacheObjectMetaRaw {
+		return false, nil // no-op
+	}
+	if !hasUnstructuredLocalObjectMetaRaw {
+		unstructured.RemoveNestedField(unstructuredCacheObject.Object, "metadata")
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:114]: is unstructuredCacheObject from a lister? We cannot touch it, not even temporarily.

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+					err = unstructured.SetNestedField(unstructuredCacheObject.Object, shardName, "metadata", "annotations", genericrequest.AnnotationKey)
+				}
+			}()
+		}
+		// TODO: in the future the original RV will be stored in an annotation
+	}
+
+	// before we can compare with the local object we need to
+	// store, remove and then bring back the ResourceVersion on the local object
+	if unstructuredLocalObjectResourceVersion, hasUnstructuredLocalObjectResourceVersion := unstructuredLocalObjectMeta["resourceVersion"]; hasUnstructuredLocalObjectResourceVersion {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:146]: localRV, found

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 30]
```diff
@@ -0,0 +1,213 @@
...
+	newCacheObjMeta := map[string]interface{}{}
+	for k, v := range localObjMeta {
+		newCacheObjMeta[k] = v
+	}
+	return true, unstructured.SetNestedMap(unstructuredCacheObject.Object, newCacheObjMeta, "metadata")
+}
+
+// ensureUnstructuredRemaining changes unstructuredCacheObject to match unstructuredLocalObject except for the metadata field
+// returns true when the unstructuredCacheObject was updated.
+func ensureUnstructuredRemaining(unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) (bool, error) {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:169]: everything in here is unstructured. Drop the prefix in the local vars.

change the long "found"-vars into just `found`, e.g. `hasUnstructuredLocalObjectMeta -> found`.

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 31]
```diff
@@ -0,0 +1,213 @@
...
+		unstructured.RemoveNestedField(localObjMeta, "resourceVersion")
+		defer func() {
+			if err == nil {
+				localObjMeta["resourceVersion"] = localObjRV
+			}
+		}()
+	}
+
+	metaChanged = !reflect.DeepEqual(cacheObjMeta, localObjMeta)
+	if !metaChanged {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:156]:
```suggestion
	if changed = !reflect.DeepEqual(cacheObjMeta, localObjMeta); changed {
```

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 32]
```diff
@@ -0,0 +1,213 @@
...
+					err = unstructured.SetNestedField(unstructuredCacheObject.Object, shard, "metadata", "annotations", genericrequest.AnnotationKey)
+				}
+			}()
+		}
+		// TODO: in the future the original RV will be stored in an annotation
+	}
+
+	// before we can compare with the local object we need to
+	// store, remove and then bring back the ResourceVersion on the local object
+	if localObjRV, hasLocalObjRV := localObjMeta["resourceVersion"]; hasLocalObjRV {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:146]:
```suggestion
	if localRV, found := localObjMeta["resourceVersion"]; found {
```

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 33]
```diff
@@ -0,0 +1,213 @@
...
+	// store, remove and then bring back fields that are unique only to the cache object
+	if cacheObjRV, hasCacheObjRV := cacheObjMeta["resourceVersion"]; hasCacheObjRV {
+		unstructured.RemoveNestedField(cacheObjMeta, "resourceVersion")
+		defer func() {
+			if err == nil {
+				err = unstructured.SetNestedField(unstructuredCacheObject.Object, cacheObjRV, "metadata", "resourceVersion")
+			}
+		}()
+	}
+	if cacheObjAnnotationsRaw, hasCacheObjAnnotations := cacheObjMeta["annotations"]; hasCacheObjAnnotations {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:128]:
```suggestion
	if cacheAnnotation, found := cacheObjMeta["annotations"]; found {
```

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 34]
```diff
@@ -0,0 +1,213 @@
...
+		return false, nil // no-op
+	}
+	if !hasLocalObjMetaRaw {
+		unstructured.RemoveNestedField(unstructuredCacheObject.Object, "metadata")
+		return true, nil
+	}
+
+	// before we can compare the cache object we need to
+	// store, remove and then bring back fields that are unique only to the cache object
+	if cacheObjRV, hasCacheObjRV := cacheObjMeta["resourceVersion"]; hasCacheObjRV {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:120]:
```suggestion
	if cacheRV, found := cacheObjMeta["resourceVersion"]; found {
```

File: pkg/reconciler/cache/replication/replication_reconcile_unstructured.go [hunk 35]
```diff
@@ -0,0 +1,213 @@
...
+//  2. deletion of the object from the cache server
+//     happens when either of the following is true:
+//     - the unstructuredLocalObject object is nil
+//     - the unstructuredLocalObject was deleted
+//  3. modification of the object to match the original/local object
+//     happens when either of the following is true:
+//      - the unstructuredLocalObject's metadata doesn't match the unstructuredCacheObject
+//      - the unstructuredLocalObject's spec doesn't match the unstructuredCacheObject
+//      - the unstructuredLocalObject's status doesn't match the unstructuredCacheObject
+func (c *controller) reconcileUnstructuredObject(ctx context.Context, cluster logicalcluster.Name, gvr *schema.GroupVersionResource, unstructuredCacheObject *unstructured.Unstructured, unstructuredLocalObject *unstructured.Unstructured) error {
```
Comment (sttts) [inline pkg/reconciler/cache/replication/replication_reconcile_unstructured.go:47]: everything is unstructure here. Drop the unstructured prefix in all vars.
